WCFDataService/Client/System/Data/Services/Client/Binding/DataServiceCollectionOfT.cs

```diff
@@ -758,15 +758,29 @@ private void InternalLoadCollection(IEnumerable<T> items)
             Debug.Assert(!(items is DataServiceQuery), "SL Client using DSQ as items...should have been caught by ValidateIteratorParameter.");
 #endif
 
+            if (this.IsTracking && this.observer != null)
+            {
+                // Performance optimazation - disable CallectionChanged notifications while enumerating items.
+                // It will be enabled and observer will be notified with the entire bulk/page once we are done enumerating.
+                this.observer.PauseTracking(this);
+            }
+            int countBefore = this.Count;
             foreach (T item in items)
             {
+                // - The check for this.Contains(item) is very expensive - please avoid it. I see no way it can return anything else than false given how this code is used.
                 // if this is too slow, consider hashing the set
                 // or just use LoadProperties                    
-                if (!this.Contains(item))
+                //if (!this.Contains(item))
```

Comment:
I checked – it does not throw. It is one of the major performance bottlenecks doing the check for Contains()
/uffe

From: ZOE [mailto:notifications@github.com]
Sent: 5. september 2016 10:51
To: OData/odata.net <odata.net@noreply.github.com>
Cc: Lauesen.Uffe UFE <UFE@kmd.dk>; Mention <mention@noreply.github.com>
Subject: Re: [OData/odata.net] Performance: Implemented support for peeking backing field values beh… (#663)


In WCFDataService/Client/System/Data/Services/Client/Binding/DataServiceCollectionOfT.cs<https://github.com/OData/odata.net/pull/663#discussion_r77488771>:

>                  // if this is too slow, consider hashing the set

>                  // or just use LoadProperties

> -                if (!this.Contains(item))

> +                //if (!this.Contains(item))

Is this save? would it be possible that dup insert will throw?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/OData/odata.net/pull/663/files/7225c040cd337c046518c9651f57768db9fb39c5#r77488771>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKq-smnupPBJiwDzCHGxJ5orsNjHqAnVks5qm9figaJpZM4JdvF4>.

Vi gør opmærksom på, at denne e-mail kan indeholde fortrolig information. Hvis du ved en fejltagelse modtager e-mailen, beder vi dig venligst informere afsender om fejlen ved at bruge svarfunktionen. Samtidig beder vi dig slette e-mailen i dit system uden at videresende eller kopiere den. Selvom e-mailen og ethvert vedhæftet bilag efter vores overbevisning er fri for virus og andre fejl, som kan påvirke computeren eller it-systemet, hvori den modtages og læses, åbnes den på modtagerens eget ansvar. Vi påtager os ikke noget ansvar for tab og skade, som er opstået i forbindelse med at modtage og bruge e-mailen.

Please note that this message may contain confidential information. If you have received this message by mistake, please inform the sender of the mistake by sending a reply, then delete the message from your system without making, distributing or retaining any copies of it. Although we believe that the message and any attachments are free from viruses and other errors that might affect the computer or it-system where it is received and read, the recipient opens the message at his or her own risk. We assume no responsibility for any loss or damage arising from the receipt or use of this message.